README.md
# Persona Example App

Built on the shoulder of giants

 - [Mozilla Persona](https://developer.mozilla.org/en-US/docs/Persona)
 - [CouchDB](http://couchdb.apache.org/) hosted at [Cloudant](https://cloudant.com/)
 - [nano](https://github.com/dscape/nano)
 - [Express](http://expressjs.com/)
 - [Request](https://github.com/mikeal/request)
 - [jQuery](http://jquery.com/)
 - [Bootstrap](http://twitter.github.io/bootstrap/)

## Demo

Visit [mysterious-coast-9759.herokuapp.com](http://mysterious-coast-9759.herokuapp.com/) to see the app in action.
It might take some time to load as it is running on the free dyno.

If you don't want to use your own email (as it is stored in the db temporarily) check out these Identity Providers

 - [mockmyid](https://mockmyid.com/)
 - [personatestuser](http://personatestuser.org/)
 
## Other examples

Take a look at these other resources. They were the foundation and inspiration for my own project. Without them I couldn't have done it! Therefore, thank you!

 - [123done](https://github.com/mozilla/123done)
 - [node.js-persona-example](https://github.com/lloyd/node.js-persona-example)
 - [myfavoritebeer.org](https://github.com/lloyd/myfavoritebeer.org)
 - [browserid-cookbook](https://github.com/mozilla/browserid-cookbook/tree/master/node-express)

However, all of the examples were missing some features I'd like to see. These are

 - proper DB implementation
 - CSRF protection
 - Content Security Policy (CSP)
 - avoid global variables
 - use of request module (imho makes code easier to read)
 - simplicity

I tried to take the best features from the examples to create my own app.

I'm sure my example can be improved as well. So if there is anything wrong or you see room for improvements, please open an [issue](https://github.com/zeMirco/mozilla-persona-express-couchdb/issues).

## More

To read more about this topic see my post [Mozilla Persona example app with Express and Couchdb](http://mircozeiss.com/mozilla-persona-example-app-with-express-and-couchdb/)
 
## Test

Coming soon ...

## License

Copyright (C) 2013 [Mirco Zeiss](mailto: mirco.zeiss@gmail.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
# 2013-05-13

 - don't use email as db key

# 2013-05-01

 - initial release

app.js

/**
 * Module dependencies.
 */

var express = require('express')
  , http = require('http')
  , path = require('path');

// routes
var index = require('./routes/index.js');
var profile = require('./routes/profile.js');
var auth = require('./routes/auth.js');

var app = express();

// csp middleware
// Important: twitter and ghbtns are not needed!
// script-src 'unsafe-inline' is needed for twitter and bootstrap
// they are here for the demo
// style-src 'unsafe-inline' is required for chrome+jquery bug
var policy =  "default-src 'self';" +
              "frame-src 'self' https://login.persona.org https://platform.twitter.com http://platform.twitter.com http://ghbtns.com;" +
              "script-src 'self' 'unsafe-inline' https://login.persona.org http://platform.twitter.com;" +
              "style-src 'self' 'unsafe-inline'";

app.use(function(req, res, next) {
  // Firefox and Internet Explorer
  res.header("X-Content-Security-Policy", policy);
  // Safari and Chrome
  res.header("X-WebKit-CSP", policy);
  // continue with next middleware
  next();
});

// all environments
app.set('port', process.env.PORT || 3000);
app.set('views', __dirname + '/views');
app.set('view engine', 'jade');
app.use(express.favicon());
app.use(express.logger('dev'));
app.use(express.bodyParser());
app.use(express.methodOverride());
app.use(express.cookieParser('your secret here'));
app.use(express.cookieSession());
app.use(express.csrf());

// custom middleware
app.use(function(req, res, next) {
  // csrf
  res.locals.token = req.session._csrf;
  // cookie
  if (req.session.email) {
    res.cookie('email', req.session.email);
  }
  // continue with router
  next();
});

app.use(app.router);
app.use(express.static(path.join(__dirname, 'public')));

// middleware to restrict access to internal routes
function restrict(req, res, next) {
  if (req.session.email) {
    next();
  } else {
    res.redirect('/');
  }
}

// set username to null for all routes
// means we only have to pass it to our views when we actually have a username
app.locals.username = null;

// development only
if ('development' == app.get('env')) {
  app.use(express.errorHandler());
}

// routes
app.get('/', index.get);
app.post('/auth/login', auth.login);
app.get('/auth/logout', auth.logout);
app.get('/profile', restrict, profile.get);
app.post('/profile', restrict, profile.post);

http.createServer(app).listen(app.get('port'), function(){
  console.log('Express server listening on port ' + app.get('port'));
});


routes/profile.js

var db = require('../db.js');

/**
 * GET /profile
 *
 * Identify user via session and get information from db.
 */
exports.get = function(req, res) {
  var email = req.session.email;

  db.view('users', 'byEmail', {key: email}, function(err, body) {
    res.render('profile', {
      title: 'Your profile',
      email: req.session.email,
      username: (body.rows[0] && body.rows[0].value.username)
    })
  })

};

/**
 * POST /profile
 *
 * Handle post request and save username to db.
 */
exports.post = function(req, res) {
  var email = req.session.email;
  var username = req.body.username;

  // get document from db
  db.view('users', 'byEmail', {key: email}, function(err, body) {
    if (err) console.log(err);
    // set username property
    var doc = body.rows[0].value;
    doc.username= username;
    // save document to db
    db.insert(doc, function(err, body) {
      res.render('profile', {
        title: 'Your profile',
        email: email,
        username: username
      })
    })
  });

};

public/javascripts/script.js
$(function() {

  $('#login').on('submit', function(event) {
    event.preventDefault();
    navigator.id.request();
  });

  $('#signout').on('click', function(event) {
    event.preventDefault();
    navigator.id.logout();
  });

  var email = $.cookie('email') || null;

  // Mozilla Persona integration
  navigator.id.watch({
    loggedInUser: email,
    onlogin: function(assertion) {
      var token = $('#token').val();
      $.post('/auth/login', {assertion: assertion, _csrf: token}, function(data) {
        window.location.reload();
      });
    },
    onlogout: function() {
      window.location = '/auth/logout';
    }
  });
  
  // twitter
  !function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');
});

public/javascripts/jquery.cookie.js
/*!
 * jQuery Cookie Plugin v1.3.1
 * https://github.com/carhartl/jquery-cookie
 *
 * Copyright 2013 Klaus Hartl
 * Released under the MIT license
 */
(function (factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as anonymous module.
    define(['jquery'], factory);
  } else {
    // Browser globals.
    factory(jQuery);
  }
}(function ($) {

  var pluses = /\+/g;

  function raw(s) {
    return s;
  }

  function decoded(s) {
    return decodeURIComponent(s.replace(pluses, ' '));
  }

  function converted(s) {
    if (s.indexOf('"') === 0) {
      // This is a quoted cookie as according to RFC2068, unescape
      s = s.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');
    }
    try {
      return config.json ? JSON.parse(s) : s;
    } catch(er) {}
  }

  var config = $.cookie = function (key, value, options) {

    // write
    if (value !== undefined) {
      options = $.extend({}, config.defaults, options);

      if (typeof options.expires === 'number') {
        var days = options.expires, t = options.expires = new Date();
        t.setDate(t.getDate() + days);
      }

      value = config.json ? JSON.stringify(value) : String(value);

      return (document.cookie = [
        config.raw ? key : encodeURIComponent(key),
        '=',
        config.raw ? value : encodeURIComponent(value),
        options.expires ? '; expires=' + options.expires.toUTCString() : '', // use expires attribute, max-age is not supported by IE
        options.path    ? '; path=' + options.path : '',
        options.domain  ? '; domain=' + options.domain : '',
        options.secure  ? '; secure' : ''
      ].join(''));
    }

    // read
    var decode = config.raw ? raw : decoded;
    var cookies = document.cookie.split('; ');
    var result = key ? undefined : {};
    for (var i = 0, l = cookies.length; i < l; i++) {
      var parts = cookies[i].split('=');
      var name = decode(parts.shift());
      var cookie = decode(parts.join('='));

      if (key && key === name) {
        result = converted(cookie);
        break;
      }

      if (!key) {
        result[name] = converted(cookie);
      }
    }

    return result;
  };

  config.defaults = {};

  $.removeCookie = function (key, options) {
    if ($.cookie(key) !== undefined) {
      // Must not alter options, thus extending a fresh object...
      $.cookie(key, '', $.extend({}, options, { expires: -1 }));
      return true;
    }
    return false;
  };

}));

routes/auth.js

var request = require('request');
var db = require('../db.js');

/**
 * POST /auth/login
 *
 * Route is part of Persona integration. It is called by navigator.id.watch -> onlogin.
 * Take the assertion sent from the client and verify against Persona Verifier.
 * If assertion is valid store email in session  and check if user is already in db.
 * If user isn't in db insert a new user document with an empty username property.
 * If user is already in db just send 200 - OK.
 */
exports.login = function(req, resp) {
  
  var audience;
  if (process.env.NODE_ENV === 'production') {
    audience = 'http://mysterious-coast-9759.herokuapp.com/'
  } else {
    audience = 'http://localhost:3000/'
  }
  
  request.post({
    url: 'https://verifier.login.persona.org/verify',
    json: {
      assertion: req.body.assertion,
      audience: audience
    }
  }, function(err, res, body) {
    if (err) console.log(err);
    if (body.status === 'okay') {
      var email = body.email;
      req.session.email = email;
      // check if email is already saved in db
      db.view('users', 'byEmail', {key: email}, function(err, body) {
        if (err) console.log(err);
        if (!body.rows[0]) {
          // email is not in db
          var user = {
            type: 'user',
            email: email,
            username: ''
          };
          db.insert(user, function(err, body) {
            if (err) console.log(err);
            resp.send(200);
          })
        } else {
          // email is already stored in db
          resp.send(200);
        }
      });
    }
  });

};

/**
 * GET /auth/logout
 *
 * Route is part of Persona integration. It is called by navigator.id.watch -> onlogout.
 * Destroy session, delete cookie and redirect user to index page.
 */
exports.logout = function(req, res) {
  req.session = null;
  res.clearCookie('email');
  res.redirect('/');
};